### PR TITLE
RFC 0012 - Invalid Values

### LINKED ISSUE #291: [RFC] Invalid Value Concept
## Short Description
Right now, we just drop invalid values. We will need a concept to allow more fine-grained handling of invalid values.

## Definition of Done
- [ ] The RFC document is created following the existing pattern
- [ ] A pull request with the RFC document is opened
- [ ] The pull request passed peer review and is accepted
- [ ] An issue is opened that tracks the implementation of the RFC

**Note:** The details of the RFC will be discussed in the PR, not in this issue!

### PR DESCRIPTION
**Pull request by georg-schwarz**
Closes #291

### CODE REVIEW

File: rfc/0012-invalid-values/README.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+
+#### Meta Table Linked
+An additional table is created with the naming convention `<value table name>_META`. In this table, there are columns that mirror the structure of the original table. For each column in the value table, we introduce three meta columns:
+- `<value column name>_isValid`, flag whether the value in the value column is valid or not, 
+- `<value column name>_details`, text for describing how the invalid value emerged,
+- `<value column name>_lastValidValue`, the last valid value before turning invalid. 
+
+Invalid values in the value table are assigned to `null`. As an implication, all value columns must be `nullable`.
+
+In the future, we might add more meta columns in this manner to the meta table.
```
Comment (felix-oq) [inline rfc/0012-invalid-values/README.md:87]: These two sections seem to be identical (except for the headline)

File: rfc/0012-invalid-values/README.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+
+#### Ignore Row
+This configuration ignores rows with invalid values. They are not written into the database. Ignored values are logged on the info level (not warning since actively configured).
+
+This resembles the current configuration.
+
+
+## Drawbacks
+- The `"meta-column"` configuration of the `SQLiteLoaderSQLiteLoader` is not really extensible for further meta data.
```
Comment (felix-oq) [inline rfc/0012-invalid-values/README.md:96]:
```suggestion
- The `"meta-column"` configuration of the `SQLiteLoader` is not really extensible for further meta data.
```

File: rfc/0012-invalid-values/README.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+## Alternatives
+- Don't introduce a concept for invalid values at all.
+- Move handling of invalid values to value type concept instead of using blocks.
```
Comment (felix-oq) [inline rfc/0012-invalid-values/README.md:100]: Here, I could think of a "invalid value safety" concept in our type system, similar to to "null safety" in modern programming languages like [Dart](https://dart.dev/null-safety) and [Kotlin](https://kotlinlang.org/docs/null-safety.html#nullable-types-and-non-null-types). E.g. having a `?` operator to turn a type into one that is able to hold `INVALID`, like `integer?`.

File: rfc/0012-invalid-values/README.md (hunk 1 above)
Comment (georg-schwarz) [reply rfc/0012-invalid-values/README.md]: Whoops, seems like I overwrote something by copy-paste. Readded it.

File: rfc/0012-invalid-values/README.md (hunk 3 above)
Comment (georg-schwarz) [reply rfc/0012-invalid-values/README.md:103]: Added it to Possible Future Enhancements.

File: rfc/0012-invalid-values/README.md [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+```
+block MyInvalidRowFilter oftype InvalidTableRowFilter {}
+```
+
+#### Block-Type `InvalidTableCellReplacer`
+This block-type allows replacing all invalid values of a column with a valid value.
+```
+block MyInvalidTemperatureReplacer oftype InvalidTableCellReplacer {
+  column: "Temperature";  // refers to table column
+  value: 21;              // expression producing valid value
```
Comment (rhazn) [inline rfc/0012-invalid-values/README.md:51]: I like this, in the future expressions should be able to access the whole table (e.g. have a value that is the average of the column or computed from other values in the table). That can be in expression extensions though, not in this RFC.

**Comment by georg-schwarz** [comment]
@rhazn Should we close this PR for now? I'm not sure if this is currently high in priority, making this RFC potentially prone to side-effects of other RFCs.

**Comment by rhazn** [comment]
Fine with me, should have a status like `abandoned` or similar. We can refer back to this / open it again in the future.

**Comment by georg-schwarz** [comment]
Do you mean the RFC should have the status `abandoned` or the corresponding issue (if there is any)?

**Comment by rhazn** [comment]
The RFC, I would just delete an issue if there is any.

**Comment by georg-schwarz** [comment]
I decided to rather close this PR and reference it in the corresponding issue.
This should give us a hint that there was already an RFC that we can pick up in the future again.
Should lead to the same results but leave our checked-in RFCs more lightweight.

**Pull request closed without merging**